src/rgw/driver/rados/rgw_rados.cc

```diff
@@ -116,6 +116,40 @@ static string default_storage_extra_pool_suffix = "rgw.buckets.non-ec";
 static RGWObjCategory main_category = RGWObjCategory::Main;
 #define RGW_USAGE_OBJ_PREFIX "usage."
 
+static inline void read_attr(std::map<std::string, bufferlist>& attrs,
+			     const std::string& attr_name,
+			     std::string& dest,
+			     bool* found = nullptr) {
+  auto i = attrs.find(attr_name);
+  if (i != attrs.end()) {
+    dest = rgw_bl_str(i->second);
+  }
+  if (found) *found = i != attrs.end();
+}
+
+static inline void read_attr(std::map<std::string, bufferlist>& attrs,
+			     const std::string& attr_name,
+			     bufferlist& dest,
+			     bool* found = nullptr) {
+  auto i = attrs.find(attr_name);
+  if (i != attrs.end()) {
+    dest = i->second;
+  }
+  if (found) *found = i != attrs.end();
+}
+
+static inline void read_attr(std::map<std::string, bufferlist>& attrs,
+			     const std::string& attr_name,
+			     bufferlist&& dest,
```

Comment:
rvalue refs are for input parameters, where the caller is done with the variable. since this is an input/output parameter, it should be a normal lvalue reference (like the unused overload above)

it's really an element of `attrs` that gets moved-from here, so `attrs` would be the thing i'd consider passing by rvalue ref. but because we're making several `read_attr()` calls on different elements, i'd leave all of the arguments as lvalue refs, and just document that the function is destructive to attrs. naming the function `pop_attr()` would probably be enough to clarify this

Reply:
I like your `pop_attr` idea; thanks!